### DISCUSSION TITLE
When to use render! vs ScopeState.render?

### DISCUSSION THREAD

**Discussion by weilbith** [DioxusLabs/dioxus#1309] [Q&A] [answered]
Hey 👋 
I'm currently learning and trying Dioxus. Something I stumbled over and haven't understood yet is when I should use the `render` macro versus the `render` function of the components `ScopeState`. I tried to read the docs and look a little at the source code. But I'm afraid I must admit I have no clue.
In my small experiments so far they where interchangeable. The officials docs seem to also mix them. So I wonder when to use what best?
Thanks for any help to understand Dioxus better. 🙏

**Comment by ealmloff** [comment, marked as answer]
They do exactly the same thing. The render macro will only work if your scope state is called cx

**Reply by weilbith** [reply]
Oh, okay interesting. I must admit that I'm in general not a fan of abbreviations due to clean code. (some of the very few things I dislike Rust patterns).  So I was wondering: `cx` of type `ScopeState` 🤔 
So I might tend to use a different naming for my project. Good to know that then the macro will not work anymore.
And I need to level up my declarative macro knowledge. Though they can't access local variables if the caller does not provide them within the token stream. 🙈
